### PR TITLE
Const-prop bugfix: only add propagation inside own block for user variables

### PR DESCRIPTION
**Pull request by felix91gr**
A testing spinoff of #71298. This one only adds the const-prop for locals that are user variables.

### CODE REVIEW

**Comment by felix91gr** [comment]
AHH I forgot to ask for a specific reviewer. Sorry @ecstatic-morse

**Comment by felix91gr** [comment]
CC @oli-obk

**Comment by wesleywiser** [comment]
Thanks! I'm just waiting for CI to turn green before starting the perf run.

**Comment by felix91gr** [comment]
Is it normal that the timer takes this long? Yesterday it took it 2.5 hours between the "Queued f498ccd with parent 0612568, future comparison URL" comment and the "Finished benchmarking try commit" comment.

**Comment by felix91gr** [comment]
It appears to be advancing, though! https://perf.rust-lang.org/status.html It was on `style-servo` not more than 15 minutes ago :)

**Comment by felix91gr** [comment]
OH BOY this is a bad one. The regressions in performance seem to be here. They lie mostly on memory performance (so `max-rss` and `faults`) but `cycles` also seems to be regressing.

I have a couple theories. My main one is that the clone of `locals_of_current_block` is causing a lot of unnecessary allocations, which mess with the rest of the program. It can definitely be avoided. I will make it compile, and upload it here.

My other theories are:
1. That maybe [this](https://github.com/rust-lang/rust/pull/71518/files#diff-9e103702275cbef342c2decd3395bf3bR906) is not the line in which to be adding this `Local` to the list of "locals of current block". This is mostly because I still not understand 100% what is the visitor doing and in what order.
2. Maybe [this call to `get_const`](https://github.com/rust-lang/rust/pull/71518/files#diff-9e103702275cbef342c2decd3395bf3bR891) is not what we should be doing. I mean it this way: we're deleting the constant from the pool as soon as we hit a `Terminator` and find it in the `BitSet` of `locals_of_current_block`. We do this by [calling `remove_const`](https://github.com/rust-lang/rust/pull/71518/files#diff-9e103702275cbef342c2decd3395bf3bR1040). Maybe instead of changing the `LocalState` of the `InterpCx` associated with the `ConstPropagator`, we could just have a flag per Local inside of `ConstPropagator` that we turn off when we've "deleted" that `Local` from our constant propagation pool. Did that make sense?

@wesleywiser @oli-obk

**Comment by felix91gr** [comment]
Removed the `.clone()` call. If it all goes green, can we get another perf run? :)

**Comment by wesleywiser** [comment]
Sorry, I probably should have explained a bit more in the original PR. Generally, we just look at the instructions metric because the others have high variance even on the most innocuous of changes. The perf results here look ok to me. I suspect it's the combination of this PR with the other PR that is causing the issue.

**Comment by felix91gr** [comment]
> Queued [845e3f5](https://github.com/rust-lang/rust/commit/845e3f5d8ad7afbe9b98ff3588f44ddceaf379bf) with parent [40008dc](https://github.com/rust-lang/rust/commit/40008dcb494a00571123b7d59115068a200ebe34), future [comparison URL](https://perf.rust-lang.org/compare.html?start=40008dcb494a00571123b7d59115068a200ebe34&end=845e3f5d8ad7afbe9b98ff3588f44ddceaf379bf).

Dangit, look at that regression! What happened? Btw [here](https://perf.rust-lang.org/compare.html?start=061256835886be38fcb2255d585ec4e1292b77d0&end=845e3f5d8ad7afbe9b98ff3588f44ddceaf379bf&stat=instructions%3Au) it is when compared to the same `0612568` parent as the prior commit.

No but seriously, what happened? I'm guessing that either the `Local::from` call or the `BitSet::remove` calls are too heavy and it shows. I'll devise another way that uses `BitSet::clear` again but without `clone` calls. There's surely a way.

**Comment by felix91gr** [comment]
Cont: yep, [`Bitset::remove`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_index/bit_set.rs.html#129-137) is pretty heavy. It uses assertions and stuff. I guess that makes sense, after all, it's receiving an index.

**Comment by felix91gr** [comment]
> (...) I'll devise another way that uses `BitSet::clear` again but without `clone` calls. There's surely a way.

I was just testing a new approach that does not call `remove_const` and uses another `BitSet` to keep track of already "forgotten" locals. In `get_const`, first we check if the `Local` belongs to the list of "forgotten ones", and return `None` if they do.

The advantage of this is that after visiting a `Terminator`, we call `BitSet::union` with the `locals_of_current_block` to save those, and then we call `clear` on the bitset that remembers the current block. This needs no cloning, no `remove_const` calls, and no fors. There's one `assert`* on `union` and the rest of the operations are just bitwise ops.

The disadvantage is that, apart from a check in `get_const` that we have to add and whose impact should be minimal... is that `get_const` is not the only way to access `LocalValue`s. A test just failed on me and made me realize this fact.

This approach of gating with `BitSet`s might not be sustainable, if we have to work with every function that wants to access a `LocalValue`. How many are there? I'll keep looking for them, but maybe this is a signal that the two `BitSet` technique isn't the path forward. and that we need to just clone the one we're using rn.

---

\* Small digression: that assert checks whether or not the two `BitSet` have the same size. It will definitely dissappear after const generics is ready, and that will be so cool :sparkles:

**Comment by wesleywiser** [comment]
I'll let @oli-obk weigh in but I think we would probably be ok merging the original version of this PR (ie, 8facb48b9efb759cb4055ac3205a400ba5f4142e and parents) since it's pretty performance neutral (some small wins, some small losses) across the board. You're welcome to give that scheme a try and see if it improves things but `BitSet::clear()` seems very simple and I doubt is responsible for much overhead so adding additional tracking to avoid calling it probably won't pay off.

File: src/test/run-fail/divide-by-zero.rs
Comment (wesleywiser) [inline src/test/run-fail/divide-by-zero.rs:2]: @oli-obk This is fine that we now emit this extra lint right?

Comment (felix91gr) [reply src/test/run-fail/divide-by-zero.rs:2]: It's there because that test and others were panicking at compile time after the new optimization. The compiler was able to deduce the panic before running the programs :D

Comment (felix91gr) [reply src/test/run-fail/divide-by-zero.rs:2]: If I understood that directive correctly, it basically says "hey rustc, it's fine, I know this code will panic. Leave it as-is, please. I wanna see it implode."

File: src/test/mir-opt/const_prop/usize_literal_propagation.rs
Comment (oli-obk) [inline src/test/mir-opt/const_prop/usize_literal_propagation.rs:1]: If you use `u32` (or some other platform independent size) instead of `usize`, you don't need this flag

File: src/test/mir-opt/simplify_match/rustc.main.SimplifyBranches-after-copy-prop.diff
Comment (oli-obk) [inline src/test/mir-opt/simplify_match/rustc.main.SimplifyBranches-after-copy-prop.diff:25]: Since `CopyProp` doesn't actually do anything here, please change it so that we see the const prop diff here

File: src/test/run-fail/divide-by-zero.rs
Comment (oli-obk) [reply src/test/run-fail/divide-by-zero.rs:2]: :rofl: it's fine (ok, no, it's absolutely great!), but please also copy all three of these tests into a single `ui` test so we show how the warning looks.

Maybe also copy these tests to mir-opt and show the diff on const prop (you can merge these into one mir-opt test by having each of the tests in a separate funciton)

Comment (felix91gr) [reply src/test/run-fail/divide-by-zero.rs:2]: > but please also copy all three of these tests into a single ui test so we show how the warning looks.

Just so I know I got you correctly: you mean a test that lets the compiler complain about these failures, right? So like, the same as these but without the `#[allow(unconditional_panic)] directive :)

> Maybe also copy these tests to mir-opt and show the diff on const prop

YAS. :rainbow: 

Speaking of mir-opt tests, I'm planning on making a test (maybe test**s**) that showcases where we're at regarding const-prop, with comments saying what the results should ideally be if const-prop was complete.

File: src/test/mir-opt/const_prop/usize_literal_propagation.rs
Comment (felix91gr) [reply src/test/mir-opt/const_prop/usize_literal_propagation.rs:1]: Indeed! Since we're not doing a `transmute` call in the test anymore, the `usize` should not be needed right? :)

File: src/test/mir-opt/simplify_match/rustc.main.SimplifyBranches-after-copy-prop.diff
Comment (felix91gr) [reply src/test/mir-opt/simplify_match/rustc.main.SimplifyBranches-after-copy-prop.diff:25]: I just did. Should I also move it to the `const_prop` folder?

File: src/test/run-fail/divide-by-zero.rs
Comment (felix91gr) [reply src/test/run-fail/divide-by-zero.rs:2]: > Maybe also copy these tests to mir-opt and show the diff on const prop

I just did this and, oh boi. Seeing how the `assert!`s that check for invalid operations, in MIR, get a `false` const right where the assertion goes, gives a certain satisfaction that is hard to describe :heart:

**Comment by felix91gr** [comment]
Rebased to fit in with the latest changes to MIR output

**Comment by felix91gr** [comment]
This is mostly ready, I think. I still want to add a medium-sized test that showcases how far const-prop has gone, and what things it still lacks. But that can be done on a different PR. The goal of that test is mostly documentation.

Anything that might be missing?

File: src/test/mir-opt/const_prop/usize_literal_propagation/32bit/rustc.main.ConstProp.diff
Comment (wesleywiser) [inline src/test/mir-opt/const_prop/usize_literal_propagation/32bit/rustc.main.ConstProp.diff:1]: I think this file can be removed right since the original test was renamed?

File: src/test/mir-opt/const_prop/usize_literal_propagation/64bit/rustc.main.ConstProp.diff
Comment (wesleywiser) [inline src/test/mir-opt/const_prop/usize_literal_propagation/64bit/rustc.main.ConstProp.diff:1]: and this file as well.

File: src/test/compile-fail/divide-by-zero.rs
Comment (wesleywiser) [inline src/test/compile-fail/divide-by-zero.rs:1]: You can move this test under `src/test/ui`

File: src/test/compile-fail/divide-by-zero.rs
Comment (wesleywiser) [inline src/test/compile-fail/divide-by-zero.rs:7]: and add a `//~ ERROR: attempt to divide by zero` here

File: src/test/compile-fail/dst-raw-slice.rs
Comment (wesleywiser) [inline src/test/compile-fail/dst-raw-slice.rs:1]: Same with the other compile-fail tests.

**Review by wesleywiser — CHANGES_REQUESTED** [review]

File: src/test/mir-opt/const_prop/usize_literal_propagation/32bit/rustc.main.ConstProp.diff
Comment (felix91gr) [reply src/test/mir-opt/const_prop/usize_literal_propagation/32bit/rustc.main.ConstProp.diff:1]: Omg, git got me again. Patches do not commute 🤦🏻

I had deleted this in the last changes and then decided to rebase them up the tree so that we had the testing changes all sort of together. But then git decided that this change was reverted. 🤦🏻

Thanks for the sanity check!

File: src/test/compile-fail/divide-by-zero.rs
Comment (felix91gr) [reply src/test/compile-fail/divide-by-zero.rs:7]: Does that have an effect outside of ui tests?

File: src/test/compile-fail/divide-by-zero.rs
Comment (felix91gr) [reply src/test/compile-fail/divide-by-zero.rs:1]: I did! I put them in `compile-fail` and in `ui/mir`. Here we check just that they always fail to compile. There we check how the ui should look.

Comment (wesleywiser) [reply src/test/compile-fail/divide-by-zero.rs:1]: The `compile-fail` tests are being migrated into ui tests. We shouldn't be adding more `compile-fail` tests unless we have to. (UI tests can check that code fails to compile as well. You just need to include `//~ ERROR` annotations)

Comment (felix91gr) [reply src/test/compile-fail/divide-by-zero.rs:1]: Ahh! Excellent.

**Comment by felix91gr** [comment]
Oh. Of course, the OOB test has to fail because it uses indices, which are `usize`. Kay, brb, fixing that.

**Review by wesleywiser — APPROVED** [review]
Wonderful! Thanks so much for working on this 👍

**Comment by felix91gr** [comment]
⛵️

**Pull request merged**
